bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java

```diff
@@ -14,55 +13,103 @@
  * KIND, either express or implied.  See the License for the
  * specific language governing permissions and limitations
  * under the License.
- **/
+ */
 package org.ballerinalang.jvm;
 
-import java.util.concurrent.Callable;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
+import org.ballerinalang.jvm.values.FutureValue;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
 import java.util.function.Function;
 
 /**
- * Simple scheduler for JBallerina.
+ * Strand scheduler for JBallerina.
  *
  * @since 0.995.0
  */
 public class Scheduler {
 
-    public static ExecutorService executorService = Executors.newFixedThreadPool(3);;
+    private LinkedList<SchedulerItem> runnableList = new LinkedList<>();
+    private Map<Strand, ArrayList<SchedulerItem>> blockedList = new HashMap<>();
 
-    public static Strand schedule(Object[] params, Function function) {
-        Strand strand = new Strand();
-        AsyncExecutor async = new AsyncExecutor(function, params);
-        strand.future = executorService.submit(async);
+    private static Scheduler scheduler = null;
 
-        return strand;
+    /**
+     * Add a task to the runnable list, which will eventually be executed by the Scheduler.
+     * @param params - parameters to be passed to the function
+     * @param function - function to be executed
+     * @return - Reference to the scheduled task
+     */
+    public FutureValue schedule(Object[] params, Function function) {
+        FutureValue future = new FutureValue();
+        params[0] = future.strand;
+        SchedulerItem item = new SchedulerItem(function, params, future);
+        runnableList.add(item);
+        return future;
     }
 
-    public static void shutdown() {
-        executorService.shutdown();
+    /**
+     * Get an instance of the Scheduler.
+     * @return - scheduler instance
+     */
+    public static Scheduler getInstance() {
+        if (scheduler == null) {
```

Comment:
Can there be multiple scheduler instances or just one for the while runtime?

Reply:
Just one for the whole runtime.

Reply:
I also don't like the singleton schedule. how about we do `new Scheduler()` in `main` and pass it down?

Reply:
Are you suggesting to keep a reference to the scheduler within the strand?